### PR TITLE
Give CC a custom User Agent

### PR DESCRIPTION
**Pull request by Wilma456**

### CODE REVIEW

**Comment by SquidDev** [comment]
This'll only work for `http.get` and `http.post` as `http.request` doesn't go through `wrapRequest`. IMO it would be nicer to do this on the Java side: just add `connection.setRequestProperty( "user-agent", "computercraft" );` after this line:

https://github.com/dan200/ComputerCraft/blob/579f7443a8c47d71efd71da41744239b5019f320/src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java#L140

**Comment by Wilma456** [comment]
Now Moved to Java Side

File: src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java [hunk 1]
```diff
@@ -138,6 +138,7 @@ public void run()
 
             // Set headers
             connection.setRequestProperty( "accept-charset", "UTF-8" );
+            connection.setRequestProperty( "user-agent", "ComputerCraft/1.80" );
```
Comment (Guard13007) [inline src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java]: Does this overwrite a custom user-agent specified by user code? Or is this applied before custom headers?

Comment (SquidDev) [inline src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java]: Before custom headers, so one should be able to overwrite it.

Comment (Wilma456) [inline src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java]: I have tested it before I had made the Commit and it works great.

Comment (Restioson) [inline src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java]: Just need to be wary about updating the version number of ComputerCraft in the useragent if we include it there...

Comment (Wilma456) [inline src/main/java/dan200/computercraft/core/apis/http/HTTPRequest.java]: Of Course

**Comment by SquidDev** [comment]
One could use `${version}` like `ComputerCraft.java`. That way it'll update automatically.

**Comment by Wilma456** [comment]
@SquidDev Crazy. I had the same idea and testing it, while you wrote this.

**Comment by Wilma456** [comment]
I tested it and it does not work. This is the User Agent:
```
ComputerCraft/${version}
```
To test it, I have downloaded a Page with this source code with wget from localhost:
```
<?php
echo $_SERVER['HTTP_USER_AGENT'];
?>
```

**Comment by SquidDev** [comment]
@Wilma456 How did you run Minecraft? It'll only substitute everything if run via Gradle.

**Comment by Wilma456** [comment]
With IntelJ Run>Minecraft Client

**Comment by Restioson** [comment]
You could run it via a gradle run configuration like `gradlew runClient`

**Comment by apemanzilla** [comment]
Why don't you just use `"ComputerCraft/" + ComputerCraft.getVersion()`?

**Comment by Wojbie** [comment]
I think it makes more sense to hardcore something that will never change in same jar file? it does skip one concatenation and a function call per http request that way.

**Comment by Lignum** [comment]
> Why don't you just use "ComputerCraft/" + ComputerCraft.getVersion()?

That just returns `"${version}"`, so it's the same thing.

**Comment by apemanzilla** [comment]
I'd argue that using a method call is better practice than using Gradle to replace strings at compile-time.

**Comment by SquidDev** [comment]
It's worth noting that template substitution is usei elsewhere:

https://github.com/dan200/ComputerCraft/blob/579f7443a8c47d71efd71da41744239b5019f320/src/main/java/dan200/computercraft/shared/computer/core/ServerComputer.java#L327

A couple of alternatives:
 - Have `public static final VERSION = ...` and use that in strings. This does introduce another way of getting the version though.
 - Use the host string as the user agent. This does mean the user agent doesn't follow the standald format however.

**Comment by Wilma456** [comment]
I had changed it to ${version} because that's a little bit faster than call a function.

 I think _HOST as User Agent is a bad Idea, because it's not standard.

**Comment by apemanzilla** [comment]
The overhead of calling `ComputerCraft.getVersion` would be negligible. IMO we should do something like what Squid suggested, with a constant for the version somewhere, to keep relying on text replacements to a minimum.

**Comment by Guard13007** [comment]
I think the overhead argument is meaningless here relative to consistency
with the codebase, which someone else pointed out substitutions of that
nature are already everywhere.

**Comment by Lignum** [comment]
> I think the overhead argument is meaningless here relative to consistency with the codebase, which someone else pointed out substitutions of that nature are already everywhere.

Yes, those kinds of microoptimisations shouldn't even be argued about in 2017. The JVM would undoubtedly inline the function call, making it boil down to the same thing anyway.

**Comment by Restioson** [comment]
Although @SquidDev's alternative suggestions do make sense, I feel like they are out of scope for this PR. The current standard should be used

**Comment by dan200** [comment]
This seems like it has the potential to break things, and CC has worked perfectly fine with it's current user agent for 5 years. Too risky and no benefit. Users are free to set their own user agent as a request property.

**Pull request closed without merging**